### PR TITLE
DAST and XML Compatability

### PR DESCRIPTION
**Pull request by siefkenj**
This PR 
 * updates the Lezer grammar to support all XML features, so valid XML documents should parse as DoenetML documents.
 * Introduces DAST (DoenetML Abstract Syntax Tree) as a format for interacting with DoenetML source code. The old `parser.ts` is still around, but it should be reworked to use DAST.
   - DAST has a `type: "error"` node that contains information about parse errors.
   - the `toXml` function can output proper XML or DoenetML-style markup (e.g., with JSX-style properties). It can also output explicit `<_error>` nodes where errors were encountered, or it can silently omit them.
 * A basic pretty-printer based off of Prettier. Lots more work needs to be done on the pretty printer to make sure it satisfies all the edge cases.
 * Add a `showFormatter` attribute to `CodeEditor` that will allow for pretty printing of code and for converting between XML and DoenetML

### CODE REVIEW

File: packages/parser/test/fragments/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Fragments of DoenetML code to test the parser.
+
+Fragments `xml-*` come from the `lezer-xml` test cases: https://github.com/lezer-parser/xml/tree/main MIT license
```
Comment (dqnykamp) [inline packages/parser/test/fragments/README.md:3]: This comment seems incorrect since xml-doenet-* files don't appear to be standard xml.

Comment (siefkenj) [reply packages/parser/test/fragments/README.md:3]: Good call. Only the `xml-##` ones are from `lezer-xml`

File: packages/parser/src/lezer-to-dast/lezer-to-dast.ts [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    if (source == null) {
+        throw new Error(
+            `If you provide a SyntaxNode, you must also provide the source string`,
+        );
+    }
+    return _lezerToDast(node, source);
+}
+
+export function _lezerToDast(node: SyntaxNode, source: string): DastRoot {
+    if (typeof node === "string") {
```
Comment (dqnykamp) [inline packages/parser/src/lezer-to-dast/lezer-to-dast.ts:67]: Delete this check since already know node is a SyntaxNode? Certainly true if `_lezerToDast` is not exported outside this file.

File: packages/parser/src/lezer-to-dast/gobble-function-arguments.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    DastText,
+} from "../types";
+import { createOffsetToPositionMap } from "./lezer-to-dast-utils";
+
+/**
+ * Functions can take arguments consisting of DastNodes. The macro parser works on strings only,
+ * so functions that take DastElement children will not be fully parsed. We manually parse through
+ * and attach the arguments in this special case.
+ *
+ * **Node**: this function may mutate the input.
```
Comment (dqnykamp) [inline packages/parser/src/lezer-to-dast/gobble-function-arguments.ts:16]: **Note**

File: packages/parser/src/lezer-to-dast/lezer-to-dast-utils.ts [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    referenceNode: DastNodes,
+) {
+    if (!nodeToUpdate.position || !referenceNode.position) {
+        return;
+    }
+    const offset = referenceNode.position.start.offset;
+    const line = referenceNode.position.start.line - 1;
+    const column = referenceNode.position.start.column - 1;
+    if (!nodeToUpdate.position.start.offset != null && offset != null) {
+        nodeToUpdate.position.start.offset! += offset;
```
Comment (dqnykamp) [inline packages/parser/src/lezer-to-dast/lezer-to-dast-utils.ts:300]: Remove the `!` after `offset`.

File: packages/parser/src/lezer-to-dast/lezer-to-dast-utils.ts [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        i = entity.to - node.from;
+    }
+    ret += rawValue.slice(i);
+    return ret;
+}
+
+export function textNodeToText(node: SyntaxNode, source: string): string {
+    if (node.name !== "Text") {
+        throw new Error(
+            `Can only get the text of an "AttributeValue" type node, not "${node.name}"`,
```
Comment (dqnykamp) [inline packages/parser/src/lezer-to-dast/lezer-to-dast-utils.ts:67]: Incorrect error message.

File: packages/parser/test/dast-advanced.test.ts [hunk 6]
```diff
@@ -0,0 +1,1017 @@
...
+                  "type": "function",
+                },
+                {
+                  "type": "text",
+                  "value": " bar",
+                },
+              ]
+            `);
+
+        // Argument with a macro in it
```
Comment (dqnykamp) [inline packages/parser/test/dast-advanced.test.ts:139]: It looks like the argument has a full component in it, not a macro.

**Review by dqnykamp — APPROVED** [review]
+1
